### PR TITLE
Disable rename DLL when using win_def_file

### PR DESCRIPTION
**Pull request by vnvo2409**
We should not rename the `*.dll` in `cc_library` when `win_def_file` is specified. Part of #11600 https://github.com/bazelbuild/bazel/issues/11600#issuecomment-645234548 

cc @meteorcloudy

### CODE REVIEW

File: src/main/java/com/google/devtools/build/lib/rules/cpp/CcLibrary.java [hunk 1]
```diff
@@ -327,7 +327,11 @@ public static void init(
     if (ruleContext.getRule().getImplicitOutputsFunction() != ImplicitOutputsFunction.NONE
         || !ccCompilationOutputs.isEmpty()) {
       if (featureConfiguration.isEnabled(CppRuleClasses.TARGETS_WINDOWS)) {
-        String dllNameSuffix = CppHelper.getDLLHashSuffix(ruleContext, featureConfiguration);
+        Artifact customDefFile = common.getWinDefFile();
+        String dllNameSuffix =
+            customDefFile != null
+                ? CppHelper.getDLLHashSuffix(ruleContext, featureConfiguration)
```
Comment (meteorcloudy) [inline src/main/java/com/google/devtools/build/lib/rules/cpp/CcLibrary.java:333]: Reversed logic? And also, can you put this logic in `getDLLHashSuffix` function? It's used at other places as well.

**Review by meteorcloudy — CHANGES_REQUESTED** [review]
Nice catch! Thanks

File: src/main/java/com/google/devtools/build/lib/rules/cpp/CcLibrary.java (hunk 1 above)
Comment (vnvo2409) [reply src/main/java/com/google/devtools/build/lib/rules/cpp/CcLibrary.java]: For some reasons, I did not realize the reversed logic. Thank you for pointing it out.

File: src/main/java/com/google/devtools/build/lib/rules/cpp/CppHelper.java [hunk 2]
```diff
@@ -949,7 +949,8 @@ public static String getDLLHashSuffix(
             ruleContext.getConfiguration().getOptions().get(CppOptions.class));
     if (cppOptions.renameDLL
         && cppOptions.dynamicMode != DynamicMode.OFF
-        && featureConfiguration.isEnabled(CppRuleClasses.TARGETS_WINDOWS)) {
+        && featureConfiguration.isEnabled(CppRuleClasses.TARGETS_WINDOWS)
+        && ruleContext.isAttrDefined("win_def_file", LABEL)) {
```
Comment (vnvo2409) [inline src/main/java/com/google/devtools/build/lib/rules/cpp/CppHelper.java:953]: Is this enough for guaranteeing that `ruleContext.getPrerequisiteArtifact("win_def_file")` will not return `null`.

File: src/main/java/com/google/devtools/build/lib/rules/cpp/CppHelper.java [hunk 3]
```diff
@@ -949,7 +949,8 @@ public static String getDLLHashSuffix(
             ruleContext.getConfiguration().getOptions().get(CppOptions.class));
     if (cppOptions.renameDLL
         && cppOptions.dynamicMode != DynamicMode.OFF
-        && featureConfiguration.isEnabled(CppRuleClasses.TARGETS_WINDOWS)) {
+        && featureConfiguration.isEnabled(CppRuleClasses.TARGETS_WINDOWS)
+        && !ruleContext.isAttrDefined("win_def_file", LABEL)) {
```
Comment (vnvo2409) [inline src/main/java/com/google/devtools/build/lib/rules/cpp/CppHelper.java:953]: Is this enough for guaranteeing that `ruleContext.getPrerequisiteArtifact("win_def_file")` will not return `null` ?

File: src/main/java/com/google/devtools/build/lib/rules/cpp/CppHelper.java (hunk 2 above)
Comment (meteorcloudy) [inline src/main/java/com/google/devtools/build/lib/rules/cpp/CppHelper.java]: I don't think that's enough, you have to check
`&& (!ruleContext.isAttrDefined("win_def_file", LABEL) || ruleContext.getPrerequisiteArtifact("win_def_file") == null)`

Please also put a comment here to explain why we need it.

File: src/main/java/com/google/devtools/build/lib/rules/cpp/CppHelper.java (hunk 3 above)
Comment (meteorcloudy) [reply src/main/java/com/google/devtools/build/lib/rules/cpp/CppHelper.java:953]: I don't think that's enough, you have to check
`&& (!ruleContext.isAttrDefined("win_def_file", LABEL) || ruleContext.getPrerequisiteArtifact("win_def_file") == null)`

Please also put a comment here to explain why we need it.

**Review by meteorcloudy — CHANGES_REQUESTED** [review]

**Review by meteorcloudy — APPROVED** [review]
Thanks!

**Pull request closed without merging**